asQ/allatoncesystem.py

```diff
@@ -176,47 +179,61 @@ def check_index(self, i, index_range='slice'):
         '''
         # set valid range
         if index_range not in self.max_indices.keys():
-            raise ValueError("index_range must be one of "+" or ".join(self.max_indices.keys()))
+            raise IndexError("index_range must be one of "+" or ".join(self.max_indices.keys()))
 
         maxidx = self.max_indices[index_range]
 
         # allow for pythonic negative indices
         minidx = -maxidx
 
         if not (minidx <= i < maxidx):
-            raise ValueError(f"index {i} outside {index_range} range {maxidx}")
+            raise IndexError(f"index {i} outside {index_range} range {maxidx}")
 
-    def shift_index(self, i, from_range='slice', to_range='slice'):
+    def shift_index(self, i, cpt=None, from_range='slice', to_range='slice'):
         '''
-        Shift timestep index from one range to another, and accounts for -ve indices
-        :arg i: timestep index to shift
-        :arg from_range: range of i. Either slice or window
-        :arg to_range: range to shift i to. Either slice or window
+        Shift timestep or component index from one range to another, and accounts for -ve indices.
+
+        If cpt is None, shifts from one timestep range to another. If cpt is not None, returns index in all-at-once function of component cpt in timestep i.
+
+        :arg i: timestep index to shift.
+        :arg cpt: None or component index in timestep i to shift.
+        :arg from_range: range of i. Either slice or window.
+        :arg to_range: range to shift i to. Either slice or window. Ignored if cpt is not None.
```

Comment:
I'm not sure that this docstring is entirely clear - can you define what it means for an index to be shifted from one range to another?

Reply:
Edited docstring to add example of shifts